mirrord/layer/src/socket.rs

```diff
@@ -254,6 +266,135 @@ impl UserSocket {
     }
 }
 
+/// Holds the [`OutgoingFilter`]s set up by the user, after a little bit of checking, see
+/// [`OutgoingSelector::new`].
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub(crate) struct OutgoingSelector {
+    /// If the address from `connect` matches this, then we send the connection through the
+    /// remote pod.
+    remote: HashSet<OutgoingFilter>,
+
+    /// If the address from `connect` matches this, then we send the connection from the local app.
+    local: HashSet<OutgoingFilter>,
+}
```

Comment:
This definition does not allow for a distinction between "not specified" and "specified to be empty".

When specifying `"local": []`everything should always be remote, but I think right now I think it would make everything be local.

The selector could have just one `HashSet` and a `bool`, or an an enum with two variants with each only 1 `HashSet`.

Reply:
Went ahead and converted it to an `enum`.

Now the semantics should be:

1. user specifies nothing: `Unfiltered`;
2. user specifies remote: `Remote([...])`;
3. user specifies local: `Local([...])`;
4. user tries to specify both: ??? crash, and it goes against the schema I think;

Reply:
Now to update the docs and see if this actually works.

Reply:
Looks good, is there a validation that the user does not pass an empty list in the configuration though?

Reply:
There is a validation during initialization, where we set up our globals and stuff, but can we do it when parsing the config itself?

Reply:
Caller-crate-side validation is ok for now if it's easier. Can you share a link to where the non-empty-list validation happens?

Reply:
[layer/socket/OutgoingSelector::try_from](https://github.com/metalbear-co/mirrord/pull/1607/files#diff-bf9c5f9340f82cf6357f9ea7b2d2a735776603ed1ead3daa2cd95db99c024650R324).

And in the layer lib initialization it crashes when we convert the filters [here](https://github.com/metalbear-co/mirrord/pull/1607/files#diff-478993a1def0c54eb220a1d4dcdc0ebd61c73c3d44b05ba2d89c4fef509996e0R425).

Reply:
```sh
    ✓ agent pod created
    ✓ pod is ready
thread '<unnamed>' panicked at 'Failed setting up outgoing traffic filter!: MissingConfigValue("Outgoing traffic filter cannot be empty!")', mirrord/layer/src/lib.rs:426:14
stack backtrace:
   0: rust_begin_unwind
             at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/panicking.rs:578:5
```

Reply:
Nice, thanks! Definitely enough validation for now.